File: src/test/java/dev/nytweetdeck/timeline/TimelineQueryFactoryTest.java
package dev.nytweetdeck.timeline;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;

class TimelineQueryFactoryTest {

    private final TimelineQueryFactory factory = new TimelineQueryFactory();

    @Test
    void buildsPaginatedQueriesForEverySupportedColumn() {
        assertThat(factory.create("homeForYou", null, "cursor").variables())
                .containsEntry("cursor", "cursor")
                .containsEntry("count", 20)
                .containsEntry("includePromotedContent", false)
                .containsEntry("withCommunity", true)
                .doesNotContainKey("latestControlAvailable");
        assertThat(factory.create("homeFollowing", null, null).variables())
                .containsEntry("enableRanking", false)
                .containsEntry("requestContext", "launch");
        assertThat(factory.create("userPosts", "42", null).variables())
                .containsEntry("userId", "42")
                .containsEntry("withQuickPromoteEligibilityTweetFields", false);
        assertThat(factory.create("list", "84", null).variables())
                .containsEntry("listId", "84");
        assertThat(factory.create("history", null, null).purpose()).isEqualTo("history");
        assertThat(factory.create("trends", null, null).purpose()).isEqualTo("trends");
        assertThat(factory.create("notifications", null, null).purpose())
                .isEqualTo("notifications");
        assertThat(factory.create("search", "NyTweetDeck", null).variables())
                .containsEntry("rawQuery", "NyTweetDeck")
                .containsEntry("product", "Latest")
                .containsEntry("withGrokTranslatedBio", false)
                .containsEntry("withQuickPromoteEligibilityTweetFields", false);
    }

    @Test
    void rejectsUnknownKindAndMissingTarget() {
        assertThatThrownBy(() -> factory.create("unknown", null, null))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("未対応");
        assertThatThrownBy(() -> factory.create("userPosts", "", null))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("対象");
    }

    @Test
    void mapsTopAndLatestToTheNativeSearchProducts() {
        assertThat(factory.create("search", "NyTweetDeck", null, "top").variables())
                .containsEntry("product", "Top");
        assertThat(factory.create("search", "NyTweetDeck", null, "latest").variables())
                .containsEntry("product", "Latest");
        assertThatThrownBy(() -> factory.create("search", "NyTweetDeck", null, "popular"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("並び順");
    }
}

File: android/app/src/test/java/dev/nytweetdeck/android/xapi/TimelineQueryFactoryTest.kt
package dev.nytweetdeck.android.xapi

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Test

class TimelineQueryFactoryTest {
    @Test
    fun followingTimelineUsesUnrankedWebVariablesAndCursor() {
        val query = TimelineQueryFactory.create("homeFollowing", null, "cursor-1")

        assertEquals("homeFollowing", query.purpose)
        assertEquals(false, query.variables["enableRanking"])
        assertEquals("cursor-1", query.variables["cursor"])
    }

    @Test
    fun searchUsesLatestProductAndRequiresTarget() {
        val query = TimelineQueryFactory.create("search", "NyTweetDeck", null)

        assertEquals("Latest", query.variables["product"])
        assertEquals("NyTweetDeck", query.variables["rawQuery"])
        assertFalse(query.variables.containsKey("cursor"))
        assertThrows(IllegalArgumentException::class.java) {
            TimelineQueryFactory.create("search", " ", null)
        }
    }

    @Test
    fun unsupportedKindIsRejected() {
        assertThrows(IllegalArgumentException::class.java) {
            TimelineQueryFactory.create("unknown", null, null)
        }
    }
}
